// holdem-engine/dist/Card.js
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);

// src/Card.ts
var Card_exports = {};
__export(Card_exports, {
  CardRank: () => CardRank,
  CardSuit: () => CardSuit,
  cardToEmojiString: () => cardToEmojiString,
  cardToString: () => cardToString
});
module.exports = __toCommonJS(Card_exports);
var CardSuit = {
  Hearts: "h",
  Diamonds: "d",
  Clubs: "c",
  Spades: "s"
};
var suitEmojiMap = {
  h: "\u2665\uFE0F",
  d: "\u2666\uFE0F",
  c: "\u2663\uFE0F",
  s: "\u2660\uFE0F"
};
function cardToEmojiString(card) {
  return `${card.rank === "T" ? "10" : card.rank}${suitEmojiMap[card.suit]}`;
}
function cardToString(card) {
  return `${card.rank}${card.suit}`;
}
var CardRank = {
  Two: 2,
  Three: 3,
  Four: 4,
  Five: 5,
  Six: 6,
  Seven: 7,
  Eight: 8,
  Nine: 9,
  Ten: "T",
  Jack: "J",
  Queen: "Q",
  King: "K",
  Ace: "A"
};
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  CardRank,
  CardSuit,
  cardToEmojiString,
  cardToString
});
