// holdem-engine/dist/chunk-HM7ITFY7.mjs
import {
  cardToEmojiString
} from "./chunk-7HMSGKED.mjs";

// src/Player.ts
var PlayerAction = {
  Fold: "Fold",
  Call: "Call",
  Raise: "Raise",
  Check: "Check",
  Bet: "Bet",
  Show: "Show",
  Hide: "Hide"
};
var Player = class {
  constructor(id, name, chips, hand = [], isFolded = false, betAmount = 0) {
    this.id = "";
    this.name = "";
    this.chips = 0;
    this.hand = [];
    this.isFolded = false;
    this.betAmount = 0;
    this.seatNumber = -1;
    this.showCards = false;
    this.id = id;
    this.name = name;
    this.chips = chips;
    this.hand = hand;
    this.isFolded = isFolded;
    this.betAmount = betAmount;
  }
  updateChips(amount) {
    this.chips += amount;
  }
  addCard(card) {
    this.hand.push(card);
  }
  /**
   * set seat number
   * @param seatNumber {number}
   */
  setSeat(seatNumber) {
    this.seatNumber = seatNumber;
  }
  setBetAmount(amount) {
    this.betAmount = amount;
  }
  getHandEmojiString() {
    return this.hand.map((card) => cardToEmojiString(card));
  }
};

export {
  PlayerAction,
  Player
};
